---
title: "Expression of L1s in brain development and adulthood"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
---

## Extracting general information about L1s
```{r message=FALSE}
library(data.table)
library(stringr)
library(pheatmap)
library("venneuler")
TE_bed <- fread("/Volumes/MyPassport/annotations/human/repeatmasker/hg38_rmsk_TEtranscripts.bed", data.table = F)
colnames(TE_bed) <- c("chr", "start", "end", "id","dot", "strand", "info")
TE_bed$TE_subfamily <- sapply(str_split(TE_bed$info, ","), `[[`, 1)
TE_bed$TE_family <- sapply(str_split(TE_bed$info, ","), `[[`, 2)
TE_bed$TE_class <- sapply(str_split(TE_bed$info, ","), `[[`, 3)

TE_bed$TE_class <- gsub("\\?", "", TE_bed$TE_class)
TE_bed <- TE_bed[which(TE_bed$TE_class %in% c("LINE", "SINE", "LTR", "Retroposon")),]
L1_freq <- data.frame(table(TE_bed[which(TE_bed$TE_family == "L1"), "TE_subfamily"]))
colnames(L1_freq) <- c("TE_subfamily", "Freq")
```

# Normalize TEs using gene expression

To account for differences in sequencing depth, we normalize the TE quantification by gene expression. Specifically, using sizeFactors as calculated by DESeq2. We will use these later in the markdown
```{r message=FALSE}
library(DESeq2)
library(ggplot2)
library(ggrepel)
# Reading gene quantification ----
path <- "/Volumes/MyPassport/l1_manuscript/gene_counts/default/"
files <- list.files(path, pattern = "csv")
files <- files[which(!endsWith(files, "summary"))]
files <- files[which(startsWith(files, "HuEm") | startsWith(files, "TBI"))]
samples <- sapply(str_split(files, ".gene_count_matrix_2.csv"), `[[`, 1)
conditions <- ifelse(grepl("HuEmCtx", samples), "Developing", "TBI")

coldata <- data.frame(sample= samples,
           condition=conditions,
           file=files)

for(i in 1:length(samples)){
  file <- paste(path, files[i], sep="")
  sample <- samples[i]
  
  tmp <- fread(file, data.table=F)
  tmp <- tmp[,c(1, ncol(tmp))]
  colnames(tmp)[2] <- sample
  
  if(i == 1){
    counts <- tmp
  }else{
    counts <- merge(counts, tmp, by="Geneid", all=T)
  }
}

rownames(coldata) <- coldata$sample
coldata_dev_tbi <- coldata[which(coldata$condition %in% c("Developing", "TBI")),]
rownames(counts) <- counts$Geneid

# Gene DEA ----
gene_dds <- DESeqDataSetFromMatrix(counts[,rownames(coldata_dev_tbi)], coldata_dev_tbi, design = ~ condition)
gene_dds$condition <- relevel(gene_dds$condition, "TBI")
gene_dds <- DESeq(gene_dds)
gene_res <- lfcShrink(gene_dds, "condition_Developing_vs_TBI")
gene_vst <- varianceStabilizingTransformation(gene_dds)
```

## Read TE quantification from TEtranscripts

TE subfamily quantification using multimapping
```{r message=FALSE}
path <- "/Volumes/MyPassport/l1_manuscript/TEcounts/multiple/"
files <- list.files(path, pattern = "cntTable")
samples <- sapply(str_split(files, ".cntTable"), `[[`, 1)
conditions <- ifelse(grepl("HuEmCtx", samples), "Developing", "TBI")

coldata <- data.frame(sample= samples,
           condition=conditions,
           file=files)

for(i in 1:length(samples)){
  file <- paste(path, files[i], sep="")
  sample <- samples[i]
  
  tmp <- fread(file, data.table=F)
  tmp <- tmp[,c(1, ncol(tmp))]
  colnames(tmp)[2] <- sample
  
  if(i == 1){
    TEcounts_multi <- tmp
  }else{
    TEcounts_multi <- merge(TEcounts_multi, tmp, by="gene/TE", all=T)
  }
}

rownames(TEcounts_multi) <- TEcounts_multi$`gene/TE`
rownames(coldata) <- coldata$sample
coldata_dev_tbi <- coldata[which(coldata$condition %in% c("Developing", "TBI")),]

TEcounts_multi <- TEcounts_multi[,c("gene/TE", rownames(coldata_dev_tbi))]
```

## How many reads quantified as genes vs TEs?

Percentage of reads mapping to genes vs TEs (Suppl Figure 1a and 3a)
```{r message=FALSE}
TEcounts_norm_multi <- TEcounts_multi[,rownames(coldata_dev_tbi)]
TEcounts_norm_multi[] <- mapply('/', TEcounts_norm_multi[,names(gene_dds$sizeFactor)], gene_dds$sizeFactor)
TEcounts_norm_multi$Geneid <- rownames(TEcounts_norm_multi)

gene_te_reads <- data.frame(Gene = colSums(TEcounts_multi[which(startsWith(TEcounts_multi$`gene/TE`, "ENS")), rownames(coldata_dev_tbi)]),
                            TE = colSums(TEcounts_multi[which(!startsWith(TEcounts_multi$`gene/TE`, "ENS")), rownames(coldata_dev_tbi)]))
gene_te_reads$sample <- rownames(gene_te_reads)
gene_te_reads <- reshape2::melt(gene_te_reads)

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/November2022/plots/num_reads_genes_TEs_bulk.pdf")
ggplot(gene_te_reads, aes(x=sample, y=value, fill=variable)) + geom_bar(stat="identity") + theme_classic() + labs(x="Sample", y="Number of reads quantified", fill = "Type") + theme(axis.text.x = element_text(angle=90, hjust=1)) 
# dev.off()
```

## L1 expression versus housekeeping and highly expressed genes

Visualization of L1 subfamilies expression and some housekeeping genes (Figure 1b and 3b).

Merge with gencode v36 to get gene names (quantification was done by gene_id)
```{r message=FALSE}
annotation <- fread("/Volumes/MyPassport/annotations/human/gencode/v36/gencode.v36.annotation.ids.name.type.tab", data.table = F, header = F)
colnames(annotation) <- c("Geneid", "gene_name", "gene_type")
annotation <- annotation[,c(1,2)]

TEcounts_norm_multi <- merge(TEcounts_norm_multi, annotation, by="Geneid", all.x=TRUE)
rownames(TEcounts_norm_multi) <- TEcounts_norm_multi$Geneid

colnames(coldata)[2] <- "Condition"
coldata$Condition <- ifelse(coldata$Condition == "TBI", "Adult", coldata$Condition)

L1s <- TEcounts_norm_multi[which(grepl("L1:LINE", TEcounts_norm_multi$Geneid)),"Geneid"]
```

## Some extra information about L1 subfamilies

Here we extract information about the size and average divergence of the L1 subfamilies which we are showing in the heatmap
```{r message=FALSE}
fa.bed <- fread("/Volumes/MyPassport/annotations/human/repeatmasker/hg38.retro.fa.out", fill=T)
colnames(fa.bed) <- c("SW_score", "perc_div", "perc_del", "perc_ins", "query_seq", "pos_in_query_begin", "pos_in_query_end", "pos_in_query_left", "strand", "TE_subfamily", "class_family", "pos_in_repeat_begin", "pos_in_repeat_end", "pos_in_repeat_left", "ID")
fa.bed$length <- fa.bed$pos_in_query_end - fa.bed$pos_in_query_begin
mean_length <- aggregate(fa.bed$length, by=list(fa.bed$TE_subfamily, fa.bed$class_family), FUN=mean)
colnames(mean_length) <- c("TE_subfamily", "info", "avg_length")
mean_per_div <- aggregate(fa.bed$perc_div, by=list(fa.bed$TE_subfamily, fa.bed$class_family), FUN=mean)
colnames(mean_per_div) <- c("TE_subfamily", "info", "avg_perc_div")
te_freq <- as.data.frame(table(fa.bed$TE_subfamily))
te_freq <- merge(te_freq, fa.bed[,c("TE_subfamily", "class_family")], by.x="Var1", by.y="TE_subfamily")
colnames(te_freq) <- c("TE_subfamily", "Freq", "info")

# Subset only L1s
l1_mean_per_div <- mean_per_div[which(mean_per_div$info == "LINE/L1"),]
l1_mean_length <- mean_length[which(mean_length$info == "LINE/L1"),]
l1_freq <- unique(te_freq[which(te_freq$info == "LINE/L1"),])
perc_div_freq_gene_intersect <- merge(l1_mean_per_div[,c("TE_subfamily", "avg_perc_div")], l1_mean_length[,c("TE_subfamily", "avg_length")], by="TE_subfamily")
perc_div_freq_gene_intersect <- merge(perc_div_freq_gene_intersect, l1_freq, by="TE_subfamily")

# coldata_tbi <- coldata_dev_tbi[which(coldata_dev_tbi$condition %in% c("TBI")),]
freq_vs_expression_per_sample <- cbind(TE_subfamily=sapply(str_split(L1s, ":"), `[[`, 1), TEcounts_norm_multi[c(L1s), rownames(coldata)])
freq_vs_expression_per_sample <- merge(perc_div_freq_gene_intersect, freq_vs_expression_per_sample, by="TE_subfamily")
freq_vs_expression_per_sample <- freq_vs_expression_per_sample[, c("TE_subfamily", "avg_perc_div", "avg_length", "Freq", rownames(coldata))]
freq_vs_expression_per_sample <- merge(unique(freq_vs_expression_per_sample[, c("TE_subfamily", "avg_perc_div", "avg_length", "Freq")]), reshape2::melt(freq_vs_expression_per_sample[, c("TE_subfamily", rownames(coldata))], by=c("TE_subfamily")), by="TE_subfamily")

freq_vs_expression_per_sample$range_avg_perc_div <- cut(freq_vs_expression_per_sample$avg_perc_div,c(0,5,10,15,20,25,30,35))
freq_vs_expression_per_sample$TE_subfamily_label <- ifelse(!duplicated(freq_vs_expression_per_sample$TE_subfamily), freq_vs_expression_per_sample$TE_subfamily, NA)

```

Plot the heatmap
```{r message=FALSE}
gene_norm_counts <- merge(annotation, counts(gene_dds, normalize=T), by.x="Geneid", by.y='row.names', all.y=T)

dev_genes_housekeep <- c("MAP2", "DCX", "GAD1", "BCL11B", "CUX1", "SOX2", "PAX6", "PCNA", "ACTB", "GAPDH", "B2M", "GUSB")
dev_genes_housekeep_counts <- gene_norm_counts[which(gene_norm_counts$gene_name %in% dev_genes_housekeep),]
rownames(dev_genes_housekeep_counts) <- dev_genes_housekeep_counts$gene_name
dev_genes_housekeep_counts <- dev_genes_housekeep_counts[,-c(1,2)]
dev_genes_housekeep_counts <- dev_genes_housekeep_counts[dev_genes_housekeep,]

TEcounts_norm_multi_cp <- TEcounts_norm_multi
TEcounts_norm_multi_cp$TE_subfamily <- sapply(str_split(TEcounts_norm_multi_cp$Geneid, ":"), `[[`, 1)
TEcounts_norm_multi_cp <- merge(perc_div_freq_gene_intersect, TEcounts_norm_multi_cp, by="TE_subfamily") # here we basically subset TEcounts to just keept subfamilies present in perc_div_freq_gene_intersect (L1s)
rownames(TEcounts_norm_multi_cp) <- TEcounts_norm_multi_cp$TE_subfamily

L1s <- c("L1HS:L1:LINE", "L1PA2:L1:LINE", "L1PA3:L1:LINE", "L1PA4:L1:LINE", "L1PA5:L1:LINE", "L1PA6:L1:LINE",
         "L1PA7:L1:LINE", "L1PA8:L1:LINE", "L1PA10:L1:LINE", "L1PA11:L1:LINE", "L1PA12:L1:LINE",
         "L1PA13:L1:LINE", "L1PA14:L1:LINE", "L1PA15:L1:LINE", "L1PA15-16:L1:LINE", "L1PA17:L1:LINE", 
         "L1MEb:L1:LINE", "L1ME3E:L1:LINE", "L1MA8:L1:LINE", "L1MEf:L1:LINE", "X9_LINE:L1:LINE", "L1M3b:L1:LINE",
         "L1M3a:L1:LINE", "L1MEg2:L1:LINE", "L1MEg1:L1:LINE", "L1M2c:L1:LINE", "L1M2a1:L1:LINE", "L1M2a:L1:LINE", "L1P:L1:LINE",
         "L1P3b:L1:LINE", "L1P4b:L1:LINE", "L1P4c:L1:LINE", "L1P4d:L1:LINE", "L1P4e:L1:LINE", "L1P5:L1:LINE")


L1counts_young_old_norm_multi <- TEcounts_norm_multi_cp[which(TEcounts_norm_multi_cp$Geneid %in% L1s),]

L1s_housekeeping_genes <- rbind(L1counts_young_old_norm_multi[, rownames(coldata_dev_tbi)],
      dev_genes_housekeep_counts[,rownames(coldata_dev_tbi)])
L1s_names <- sapply(str_split(L1s, ":"), `[[`, 1)

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/November2022/plots/heatmap_L1s_housekeeping_genes.pdf", height=24, width = 10)
pheatmap(log2(L1s_housekeeping_genes[c(L1s_names, rownames(dev_genes_housekeep_counts)),coldata$sample]+0.5),cluster_rows = F, cluster_cols = F, annotation_col = coldata[,"Condition",drop=F],
         gaps_col = 6, show_colnames = T, fontsize = 15, annotation_colors = list("Condition" = c("Adult" = "#F8766D", "Developing" = "#00BFC4")), annotation_row = TEcounts_norm_multi_cp[,c("Freq", "avg_perc_div", "avg_length")])
# dev.off()
```

# Unique element quantification

Here we read the count matrices of individual TEs. We have performed the quantification in both directions - we have divided the bam files into forward and reverse transcription, and quantified both of them. 

The matrices we read here (.TE_count_matrix_per_strand.csv) contain one column with the counts using the forward-stranded bam files, and one with the reverse-stranded bam (labeled as sample_forward or sample_reverse). 

We normalize their expression using the sizeFactors of the relevant sample form our gene_dds object.
```{r message=FALSE}
# Read TE quantification ---- 
TEclass <- fread('/Volumes/MyPassport/annotations/human/repeatmasker/hg38_rmsk_TEtranscripts_classification.tab', data.table = F, header = F)
colnames(TEclass) <- c("TE_id", "TE_subfamily", "TE_family", "TE_class")

path <- "/Volumes/MyPassport/l1_manuscript/TEcounts/unique/"
files <- list.files(path, pattern = "csv")
files <- files[which(!grepl("summary", files))]
files <- files[which(!grepl("ORF0", files))]
files <- files[which(startsWith(files, "HuEm") | startsWith(files, "TBI"))]
samples <- unique(sapply(str_split(files, ".TE_count_matrix_per_strand.csv"), `[[`, 1))

for(i in 1:length(samples)){
  file <- paste(path, files[i], sep="")
  sample <- samples[i]
  
  tmp <- fread(file, data.table=F)
  tmp <- tmp[,c("Geneid", "Strand", "Length", colnames(tmp)[c(ncol(tmp), ncol(tmp)-1)])]
  colnames(tmp)[c(ncol(tmp), ncol(tmp)-1)] <- paste(sample, sapply(str_split(sapply(str_split(colnames(tmp)[c(ncol(tmp), ncol(tmp)-1)], "sortedByCoord."), `[[`, 2), ".out.bam"), `[[`, 1), sep="_")
    
  if(i == 1){
    TEcounts <- tmp
  }else{
    if(all(tmp$Geneid == TEcounts$Geneid)){ # If they are in the same order, we can use cbind so that is faster...
      TEcounts <- cbind(TEcounts, tmp[,c(ncol(tmp), (ncol(tmp)-1)),drop=F])  
    }else{
      TEcounts <- merge(TEcounts, tmp, by=c("Geneid", "Strand", "Length"), all=T)  
    }
  }
  print(sample)
}
rownames(TEcounts) <- TEcounts$Geneid

# Divide matrix by transcription direction
TEcounts_forward <- TEcounts[, c("Strand", colnames(TEcounts)[which(grepl("forward", colnames(TEcounts)))])]
TEcounts_reverse <- TEcounts[, c("Strand", colnames(TEcounts)[which(grepl("reverse", colnames(TEcounts)))])]
colnames(TEcounts_forward) <- sapply(str_split(colnames(TEcounts_forward), "_forward"), `[[`, 1)
colnames(TEcounts_reverse) <- sapply(str_split(colnames(TEcounts_reverse), "_reverse"), `[[`, 1)

TEcounts_forward_norm <- TEcounts_forward[coldata$sample]
TEcounts_reverse_norm <- TEcounts_reverse[coldata$sample]

# Normalize by gene expression
TEcounts_forward_norm[] <- mapply(TEcounts_forward_norm, gene_dds$sizeFactor[coldata$sample], FUN="/")
TEcounts_reverse_norm[] <- mapply(TEcounts_reverse_norm, gene_dds$sizeFactor[coldata$sample], FUN="/")

colnames(TEcounts_forward_norm) <- paste(colnames(TEcounts_forward_norm), "forward", sep="_")
colnames(TEcounts_reverse_norm) <- paste(colnames(TEcounts_reverse_norm), "reverse", sep="_")

# Annotate TEs
TEcounts_forward_norm <- merge(TEclass, TEcounts_forward_norm, by.x='TE_id', by.y="row.names")
TEcounts_reverse_norm <- merge(TEclass, TEcounts_reverse_norm, by.x='TE_id', by.y="row.names")
# Remove ? endings
TEcounts_forward_norm$TE_class <- as.character(TEcounts_forward_norm$TE_class)
TEcounts_reverse_norm$TE_class <- as.character(TEcounts_reverse_norm$TE_class)
TEcounts_forward_norm$TE_class <- gsub("\\?", "", TEcounts_forward_norm$TE_class)
TEcounts_reverse_norm$TE_class <- gsub("\\?", "", TEcounts_reverse_norm$TE_class)

# Subset to retrotransposons to not work with huge dataframes...
TEcounts_forward_norm <- TEcounts_forward_norm[which(gsub("?", "", TEcounts_forward_norm$TE_class) %in% c("LINE", "SINE", "LTR", "Retroposon")),]
TEcounts_reverse_norm <- TEcounts_reverse_norm[which(gsub("?", "", TEcounts_reverse_norm$TE_class) %in% c("LINE", "SINE", "LTR", "Retroposon")),]

TEcounts_norm <- merge(TEcounts_forward_norm, TEcounts_reverse_norm, by=c("TE_id", "TE_subfamily", "TE_family", "TE_class"), all=T)
TEcounts_norm <- merge(TEcounts_norm, TEcounts[,c("Geneid", "Length", "Strand")], by.x="TE_id", by.y="Geneid")
L1counts_norm <- TEcounts_norm[which(TEcounts_norm$TE_family == "L1"),]
```

## Ratio of full length L1HS-L1PA4 expressed

Here we are interested on calculating the ratio of full length L1HS-PA4 expressed in both data sets (adult and brain development) (Figure 1e and 3e). 

We decided to call "full length" (FL) to any L1 with a length over 6kbps. We define if an element is "expressed" if it has a mean expression of over 10 counts (normalized) (in sense) in the particular dataset.

To calculate the ratios, we also need the number of elements in that particular subfamily (and strand).

We start with the adult samples (colnames starts with TBI)
```{r message=FALSE}
# Define FL
FL_L1counts <- L1counts_norm[which(L1counts_norm$Length >= 6000),]

# Define expressed elements
fwd_L1_elements_adult <- data.frame(table(FL_L1counts[which(FL_L1counts$Strand == "+"),"TE_subfamily"], (rowSums(FL_L1counts[which(FL_L1counts$Strand == "+"), which(endsWith(colnames(FL_L1counts), "forward") & startsWith(colnames(FL_L1counts), "TBI"))]) > 10)))
rev_L1_elements_adult <- data.frame(table(FL_L1counts[which(FL_L1counts$Strand == "-"),"TE_subfamily"], (rowSums(FL_L1counts[which(FL_L1counts$Strand == "-"), which(endsWith(colnames(FL_L1counts), "reverse") & startsWith(colnames(FL_L1counts), "TBI"))]) > 10)))
colnames(rev_L1_elements_adult) <- c("TE_subfamily", "sense", "Freq")
colnames(fwd_L1_elements_adult) <- c("TE_subfamily", "sense", "Freq")
rev_L1_elements_adult <- rev_L1_elements_adult[which(rev_L1_elements_adult$sense == TRUE),]
fwd_L1_elements_adult <- fwd_L1_elements_adult[which(fwd_L1_elements_adult$sense == TRUE),]
rev_L1_elements_adult$sense <- "reverse"
fwd_L1_elements_adult$sense <- "forward"
num_L1_elements_adult <- rbind(fwd_L1_elements_adult,
                      rev_L1_elements_adult)
num_L1_elements_adult <- num_L1_elements_adult[which(num_L1_elements_adult$TE_subfamily %in% unique(freq_vs_expression_per_sample$TE_subfamily)),]

# Extract the total number of FL L1s on a particular strand
num_total_L1_elements_forward <- as.data.frame(table(FL_L1counts$TE_subfamily, FL_L1counts$Strand == "+"))
num_total_L1_elements_forward <- num_total_L1_elements_forward[which(num_total_L1_elements_forward$Var2 == TRUE),]
num_total_L1_elements_forward$Var2 <- "forward"
num_total_L1_elements_reverse <- as.data.frame(table(FL_L1counts$TE_subfamily, FL_L1counts$Strand == "-"))
num_total_L1_elements_reverse <- num_total_L1_elements_reverse[which(num_total_L1_elements_reverse$Var2 == TRUE),]
num_total_L1_elements_reverse$Var2 <- "reverse"

num_total_L1_elements <- rbind(num_total_L1_elements_forward, num_total_L1_elements_reverse)
colnames(num_total_L1_elements) <- c("TE_subfamily", "sense", "Freq")
num_total_L1_elements$set <- "total"
num_L1_elements_adult$set <- "expressed"

num_L1_elements_adult <- rbind(num_total_L1_elements, num_L1_elements_adult)
num_L1_elements_adult$set <- factor(num_L1_elements_adult$set, levels=c("total", "expressed"))

num_L1_elements_adult <- aggregate(num_L1_elements_adult$Freq, by=list(num_L1_elements_adult$TE_subfamily, num_L1_elements_adult$set), FUN=sum)
colnames(num_L1_elements_adult) <- c("TE_subfamily", "set", "Freq")
num_L1_elements_adult <- reshape2::dcast(num_L1_elements_adult, formula = TE_subfamily~set, value.var = "Freq")

num_L1_elements_adult$TE_subfamily <- factor(num_L1_elements_adult$TE_subfamily, levels=unique(freq_vs_expression_per_sample[order(freq_vs_expression_per_sample$avg_perc_div), "TE_subfamily"]))
num_L1_elements_adult_L1HS_L1PA4 <- num_L1_elements_adult[which(num_L1_elements_adult$TE_subfamily %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4")),]

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/November2022/plots/L1HS_L1PA4_ratio_of_expressed_expression_adult.pdf", height = 3, width = 3)
ggplot(num_L1_elements_adult_L1HS_L1PA4, aes(x=TE_subfamily, y=expressed/total)) + geom_bar(stat = "identity") + theme_classic() + labs(y="Number of elements with >10 reads mapped / total number of elements", x="TE family") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
# dev.off()

```

We repeat this process in the fetal samples (colnames starts with HuEmCtx)
```{r message=FALSE}
fwd_L1_elements_bd <- data.frame(table(FL_L1counts[which(FL_L1counts$Strand == "+"),"TE_subfamily"], (rowSums(FL_L1counts[which(FL_L1counts$Strand == "+"), which(endsWith(colnames(FL_L1counts), "forward") & startsWith(colnames(FL_L1counts), "HuEmCtx"))]) > 10)))
rev_L1_elements_bd <- data.frame(table(FL_L1counts[which(FL_L1counts$Strand == "-"),"TE_subfamily"], (rowSums(FL_L1counts[which(FL_L1counts$Strand == "-"), which(endsWith(colnames(FL_L1counts), "reverse") & startsWith(colnames(FL_L1counts), "HuEmCtx"))]) > 10)))
colnames(rev_L1_elements_bd) <- c("TE_subfamily", "sense", "Freq")
colnames(fwd_L1_elements_bd) <- c("TE_subfamily", "sense", "Freq")
rev_L1_elements_bd <- rev_L1_elements_bd[which(rev_L1_elements_bd$sense == TRUE),]
fwd_L1_elements_bd <- fwd_L1_elements_bd[which(fwd_L1_elements_bd$sense == TRUE),]
rev_L1_elements_bd$sense <- "reverse"
fwd_L1_elements_bd$sense <- "forward"
num_L1_elements_bd <- rbind(fwd_L1_elements_bd,
                      rev_L1_elements_bd)
num_L1_elements_bd <- num_L1_elements_bd[which(num_L1_elements_bd$TE_subfamily %in% unique(freq_vs_expression_per_sample$TE_subfamily)),]

num_total_L1_elements_forward <- as.data.frame(table(FL_L1counts$TE_subfamily, FL_L1counts$Strand == "+"))
num_total_L1_elements_forward <- num_total_L1_elements_forward[which(num_total_L1_elements_forward$Var2 == TRUE),]
num_total_L1_elements_forward$Var2 <- "forward"
num_total_L1_elements_reverse <- as.data.frame(table(FL_L1counts$TE_subfamily, FL_L1counts$Strand == "-"))
num_total_L1_elements_reverse <- num_total_L1_elements_reverse[which(num_total_L1_elements_reverse$Var2 == TRUE),]
num_total_L1_elements_reverse$Var2 <- "reverse"

num_total_L1_elements <- rbind(num_total_L1_elements_forward, num_total_L1_elements_reverse)
colnames(num_total_L1_elements) <- c("TE_subfamily", "sense", "Freq")
num_total_L1_elements$set <- "total"
num_L1_elements_bd$set <- "expressed"

num_L1_elements_bd <- rbind(num_total_L1_elements, num_L1_elements_bd)
num_L1_elements_bd$set <- factor(num_L1_elements_bd$set, levels=c("total", "expressed"))

num_L1_elements_bd <- aggregate(num_L1_elements_bd$Freq, by=list(num_L1_elements_bd$TE_subfamily, num_L1_elements_bd$set), FUN=sum)
colnames(num_L1_elements_bd) <- c("TE_subfamily", "set", "Freq")
num_L1_elements_bd <- reshape2::dcast(num_L1_elements_bd, formula = TE_subfamily~set, value.var = "Freq")

num_L1_elements_bd$TE_subfamily <- factor(num_L1_elements_bd$TE_subfamily, levels=unique(freq_vs_expression_per_sample[order(freq_vs_expression_per_sample$avg_perc_div), "TE_subfamily"]))
num_L1_elements_bd_L1HS_L1PA4 <- num_L1_elements_bd[which(num_L1_elements_bd$TE_subfamily %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4")),]

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/November2022/plots/L1HS_L1PA4_ratio_of_expressed_expression_bd.pdf", height = 3, width = 3)
ggplot(num_L1_elements_bd_L1HS_L1PA4, aes(x=TE_subfamily, y=expressed/total)) + geom_bar(stat = "identity") + theme_classic() + labs(y="Number of elements with >10 reads mapped / total number of elements", x="TE family") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))# + scale_fill_distiller(palette="RdYlBu")
# dev.off()

```

## How many FL L1HS-L1PA4 are expressed in each timepoint and how many are shared?

In the previous two chunks, we needed to just have the TE subfamily name to merge it to the total number of elements in that subfamily (and calculate the ratio). However, here we want to keep the unique identifier to the element that is expressed and compare it to the two datasets (Figure 4a).

We start with those expressed in our:

### RNAseq 
```{r message=FALSE}
TEcounts <- merge(TEcounts, TEclass, by.x="Geneid", by.y="TE_id")
TEcounts$TE_class <- gsub("\\?", "", TEcounts$TE_class)

# How many of these were found expressed in the brain development samples as well?
fwd_FL_L1s_adult_counts <- FL_L1counts[which(FL_L1counts$Strand == "+"), c(1,2, 3, 4, which(endsWith(colnames(FL_L1counts), "forward") & startsWith(colnames(FL_L1counts), "TBI")))]
fwd_FL_L1s_adult_counts <- fwd_FL_L1s_adult_counts[which(rowSums(fwd_FL_L1s_adult_counts[,-(4:1)]) > 10),] # Without the first three cols, as they are geneid, TE_subfamily and TE_family
rev_FL_L1s_adult_counts <- FL_L1counts[which(FL_L1counts$Strand == "-"), c(1,2, 3, 4, which(endsWith(colnames(FL_L1counts), "reverse") & startsWith(colnames(FL_L1counts), "TBI")))]
rev_FL_L1s_adult_counts <- rev_FL_L1s_adult_counts[which(rowSums(rev_FL_L1s_adult_counts[,-(4:1)]) > 10),] # Without the first three cols, as they are geneid, TE_subfamily and TE_family
# How many FL_L1s do we have expressed in the adult samples? (again, to check if we didnt make a mistake)
length(c(fwd_FL_L1s_adult_counts$TE_id, rev_FL_L1s_adult_counts$TE_id))

fwd_FL_L1s_bd_counts <- FL_L1counts[which(FL_L1counts$Strand == "+"), c(1, 2, 3, 4, which(endsWith(colnames(FL_L1counts), "forward") & startsWith(colnames(FL_L1counts), "HuEmCtx")))]
fwd_FL_L1s_bd_counts <- fwd_FL_L1s_bd_counts[which(rowSums(fwd_FL_L1s_bd_counts[,-(4:1)]) > 10),] # Without the first three cols, as they are geneid, TE_subfamily and TE_family
rev_FL_L1s_bd_counts <- FL_L1counts[which(FL_L1counts$Strand == "-"), c(1, 2, 3, 4, which(endsWith(colnames(FL_L1counts), "reverse") & startsWith(colnames(FL_L1counts), "HuEmCtx")))]
rev_FL_L1s_bd_counts <- rev_FL_L1s_bd_counts[which(rowSums(rev_FL_L1s_bd_counts[,-(4:1)]) > 10),] # Without the first three cols, as they are geneid, TE_subfamily and TE_family
# How many FL_L1s do we have expressed in the brain development samples? (again, to check if we didnt make a mistake)
length(c(fwd_FL_L1s_bd_counts$TE_id, rev_FL_L1s_bd_counts$TE_id))

fwd_L1HS_L1PA4_bd_counts <- fwd_FL_L1s_bd_counts[which(fwd_FL_L1s_bd_counts$TE_subfamily %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4")),]
rev_L1HS_L1PA4_bd_counts <- rev_FL_L1s_bd_counts[which(rev_FL_L1s_bd_counts$TE_subfamily %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4")),]
fwd_L1HS_L1PA4_adult_counts <- fwd_FL_L1s_adult_counts[which(fwd_FL_L1s_adult_counts$TE_subfamily %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4")),]
rev_L1HS_L1PA4_adult_counts <- rev_FL_L1s_adult_counts[which(rev_FL_L1s_adult_counts$TE_subfamily %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4")),]

length(c(fwd_L1HS_L1PA4_adult_counts$TE_id, rev_L1HS_L1PA4_adult_counts$TE_id))
table(c(fwd_L1HS_L1PA4_bd_counts$TE_id, rev_L1HS_L1PA4_bd_counts$TE_id) %in% c(fwd_L1HS_L1PA4_adult_counts$TE_id, rev_L1HS_L1PA4_adult_counts$TE_id))

bd_FL_L1HS_L1PA4 <- c(fwd_L1HS_L1PA4_bd_counts$TE_id, rev_L1HS_L1PA4_bd_counts$TE_id)
adult_FL_L1HS_L1PA4 <- c(fwd_L1HS_L1PA4_adult_counts$TE_id, rev_L1HS_L1PA4_adult_counts$TE_id)
bd_adult_FL_L1HS_L1PA4 <- bd_FL_L1HS_L1PA4[which(bd_FL_L1HS_L1PA4 %in% adult_FL_L1HS_L1PA4)]
bd_FL_L1HS_L1PA4 <- bd_FL_L1HS_L1PA4[which(!bd_FL_L1HS_L1PA4 %in% bd_adult_FL_L1HS_L1PA4)]
adult_FL_L1HS_L1PA4 <- adult_FL_L1HS_L1PA4[which(!adult_FL_L1HS_L1PA4 %in% bd_adult_FL_L1HS_L1PA4)]

# FL L1HS to L1PA4 (values from line 695 fetalcortex_bulk.Rmd)
# pdf("/Volumes/MyPassport/FetalCortex/manuscript/November2022/plots/venn_rnaseq_FL_L1HS_L1PA2_L1PA3_L1PA4.pdf", width = 16, height = 5)
plot(venneuler(c(Adult = length(adult_FL_L1HS_L1PA4), # Adult     
                 Development = length(bd_FL_L1HS_L1PA4), # Brain development
                 "Adult&Development" = length(bd_adult_FL_L1HS_L1PA4))))
# dev.off()
```

#### Where are these L1s? 
 
Are they intragenic or intragenic? What is the distribution of all FL L1HS-PA4? (Figure 4c)

We will use bedr to intersect our expressed FL L1HS-PA4 with the gencode v36 annotation.
```{r message=FALSE}
annotation_bed <- fread("/Volumes/MyPassport/annotations/human/gencode/v36/gencode.v36.annotation.gene.bed", data.table = F, header = F)
colnames(annotation_bed) <- c("chr", "start", "end", "gene_id", "dot", "strand")

library(bedr)
# This function just formats the coordinates into the required format for bedr. You can also add a window surrounding the region of interest (optional)
add_windows_coords <- function(df, window = 0){
  if(window > 0){
    for(i in 1:nrow(df)){
      df[i,"start"] <- df[i,"start"] - window
      df[i,"start"] <- ifelse(df[i,"start"] <= 0, 1, df[i,"start"])
      df[i,"end"] <- df[i,"end"] + window
    }
  }
  coords <- paste(df$chr, paste(df$start, df$end, sep = "-"), sep = ":")
  return(coords)
}

annotation_bed$coords <- add_windows_coords(annotation_bed)
# annotation_gtf_genes
TE_bed$chr <- ifelse(grepl("_", TE_bed$chr), sapply(str_split(TE_bed$chr, "_"), `[[`, 1), TE_bed$chr)
TE_bed$coords <- add_windows_coords(TE_bed)
rownames(TE_bed) <- TE_bed$id

genes_coords <- bedr.sort.region(unique(annotation_bed$coords))

# This function returns if a coordinate intersected with a gene (aka if its intragenic or intergenic). An element has to be 25% in the gene to be counted as intragenic
intergenic_exonic_intronic <- function(coords){
  type <- ifelse(in.region(x = coords, y = genes_coords, proportion.overlap = 0.25), "intragenic", "intergenic") # if a quarter of the TE is inside a gene - it's genic.
  return(type)
}

# This function extracts the coordinates of the genes the coordinates intersected with
TE_intragenic_genes <- function(coords){
  coords_intragenic <- bedr(
    input = list(a = coords, b = genes_coords), 
    method = "intersect", 
    params = "-u -sorted -f 0.25"
  );
  # coords[which(in.region(x = coords, y = genes_coords, proportion.overlap = 0.25))] # if a quarter of the TE is inside a gene - it's genic.
  coords_genes <- bedr(
    input = list(a = genes_coords, b = coords_intragenic), 
    method = "intersect", 
    params = "-u -sorted -F 0.25"
  );
  
  return(coords_genes)
}

bd_adult_FL_L1HS_L1PA4_coords <- bedr.sort.region(TE_bed[bd_adult_FL_L1HS_L1PA4, "coords"])
bd_adult_FL_L1HS_L1PA4_location <- TE_bed[match(bd_adult_FL_L1HS_L1PA4_coords, TE_bed$coords), c("id", "coords")]
bd_adult_FL_L1HS_L1PA4_location$location <- intergenic_exonic_intronic(bd_adult_FL_L1HS_L1PA4_location$coords)

bd_FL_L1HS_L1PA4_coords <- bedr.sort.region(TE_bed[bd_FL_L1HS_L1PA4, "coords"])
bd_FL_L1HS_L1PA4_location <- TE_bed[match(bd_FL_L1HS_L1PA4_coords, TE_bed$coords), c("id", "coords")]
bd_FL_L1HS_L1PA4_location$location <- intergenic_exonic_intronic(bd_FL_L1HS_L1PA4_location$coords)

adult_FL_L1HS_L1PA4_coords_no_chrUn <- TE_bed[which(TE_bed$id %in% adult_FL_L1HS_L1PA4 & TE_bed$chr != "chrUn"), "coords"]
adult_FL_L1HS_L1PA4_coords_no_chrUn <- bedr.sort.region(adult_FL_L1HS_L1PA4_coords_no_chrUn)
adult_FL_L1HS_L1PA4_location <- TE_bed[match(adult_FL_L1HS_L1PA4_coords_no_chrUn, TE_bed$coords), c("id", "coords")]
adult_FL_L1HS_L1PA4_location$location <- intergenic_exonic_intronic(adult_FL_L1HS_L1PA4_location$coords)
adult_FL_L1HS_L1PA4_location <- rbind(adult_FL_L1HS_L1PA4_location,
                                      cbind(TE_bed[which(TE_bed$id %in% adult_FL_L1HS_L1PA4 & TE_bed$chr == "chrUn"),c("id", "coords")], location = "chrUn"))

TE_bed$length <- TE_bed$end - TE_bed$start
all_FL_L1HS_L1PA4_coords <- data.frame(coords = bedr.sort.region(TE_bed[which(TE_bed$TE_subfamily %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4") & 
                                                            TE_bed$length >= 6000 & 
                                                            TE_bed$chr != "chrUn"), "coords"]))
all_FL_L1HS_L1PA4_coords$location <- intergenic_exonic_intronic(all_FL_L1HS_L1PA4_coords$coords)

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/intergenic_exonic_intronic_rna.pdf", height = 4, width = 3)
ggplot(bd_FL_L1HS_L1PA4_location, aes(x=location)) + geom_histogram(stat = "count") + theme_classic() + ggtitle("Brain development")
ggplot(bd_adult_FL_L1HS_L1PA4_location, aes(x=location)) + geom_histogram(stat = "count") + theme_classic() + ggtitle("Intersection")
ggplot(adult_FL_L1HS_L1PA4_location, aes(x=location)) + geom_histogram(stat = "count") + theme_classic() + ggtitle("Adult")
ggplot(all_FL_L1HS_L1PA4_coords, aes(x=location)) + geom_histogram(stat = "count") + theme_classic() + ggtitle("All >6kbp L1HS-L1PA4")
# dev.off()
```

#### Genes with intragenic L1s

Are these genes generally more or less expressed in the particular dataset which is expressing the intragenic L1? In other words, does the expression correlates with the L1? Or does it block it? 

This data is shown in Figure 4d
```{r message=FALSE}
bd_FL_L1HS_L1PA4_intragenic_genes <- annotation_bed[which(annotation_bed$coords %in% TE_intragenic_genes(bd_FL_L1HS_L1PA4_location$coords)), "gene_id"]
bd_adult_FL_L1HS_L1PA4_intragenic_genes <- annotation_bed[which(annotation_bed$coords %in% TE_intragenic_genes(bd_adult_FL_L1HS_L1PA4_location$coords)), "gene_id"]
adult_FL_L1HS_L1PA4_location_no_chrun <- adult_FL_L1HS_L1PA4_location[!startsWith(adult_FL_L1HS_L1PA4_location$coords, "chrUn"),]
adult_FL_L1HS_L1PA4_intragenic_genes <- annotation_bed[which(annotation_bed$coords %in% TE_intragenic_genes(adult_FL_L1HS_L1PA4_location_no_chrun$coords)), "gene_id"]

rownames(gene_norm_counts) <- gene_norm_counts$Geneid

gene_res_df <- as.data.frame(gene_res)
gene_res_df$type <- ifelse(rownames(gene_res_df) %in% bd_FL_L1HS_L1PA4_intragenic_genes, "Brain development",
                           ifelse(rownames(gene_res_df) %in% bd_adult_FL_L1HS_L1PA4_intragenic_genes, "Intersect",
                                  ifelse(rownames(gene_res_df) %in% adult_FL_L1HS_L1PA4_intragenic_genes, "Adult", "N.s.")))
# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/violin_genes_with_intragenic_L1.pdf", height = 3, width = 4)
ggplot(gene_res_df[which(gene_res_df$type != "N.s."),], aes(x=type, y=log2FoldChange, fill=type)) + geom_violin()  + geom_boxplot(width=0.1, color="black", alpha=0.2) + theme_classic() + geom_hline(yintercept = 0, linetype="dashed", colour="red")
# dev.off()

# Heatmaps ----
bd_FL_L1HS_L1PA4_intragenic_genes_counts <- gene_norm_counts[bd_FL_L1HS_L1PA4_intragenic_genes,]
rownames(bd_FL_L1HS_L1PA4_intragenic_genes_counts) <- bd_FL_L1HS_L1PA4_intragenic_genes_counts$gene_name
bd_adult_FL_L1HS_L1PA4_intragenic_genes_counts <- gene_norm_counts[bd_adult_FL_L1HS_L1PA4_intragenic_genes,]
rownames(bd_adult_FL_L1HS_L1PA4_intragenic_genes_counts) <- bd_adult_FL_L1HS_L1PA4_intragenic_genes_counts$gene_name
adult_FL_L1HS_L1PA4_location_no_chrun_counts <- gene_norm_counts[adult_FL_L1HS_L1PA4_intragenic_genes,]
rownames(adult_FL_L1HS_L1PA4_location_no_chrun_counts) <- adult_FL_L1HS_L1PA4_location_no_chrun_counts$gene_name

bd_FL_L1HS_L1PA4_intragenic_genes_counts <- bd_FL_L1HS_L1PA4_intragenic_genes_counts[,coldata$sample]
bd_adult_FL_L1HS_L1PA4_intragenic_genes_counts <- bd_adult_FL_L1HS_L1PA4_intragenic_genes_counts[,coldata$sample]
adult_FL_L1HS_L1PA4_location_no_chrun_counts <- adult_FL_L1HS_L1PA4_location_no_chrun_counts[,coldata$sample]

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/genes_with_intragenic_L1s.pdf")
pheatmap(log2(bd_FL_L1HS_L1PA4_intragenic_genes_counts[which(rowSums(bd_FL_L1HS_L1PA4_intragenic_genes_counts) > 0),]+0.5), cluster_rows = F, cluster_cols = F, annotation_col = coldata[,"Condition",drop=F], gaps_col = 6, show_colnames = F, show_rownames = F, fontsize = 15, annotation_colors = list("Condition" = c("Adult" = "#F8766D", "Developing" = "#00BFC4")), border_color = NA, scale = "row", main = "Genes with intragenic >6kbp L1HS-L1PA4\nBrain development specific", na_col = "black")

pheatmap(log2(bd_adult_FL_L1HS_L1PA4_intragenic_genes_counts[which(rowSums(bd_adult_FL_L1HS_L1PA4_intragenic_genes_counts) > 0),]+0.5), cluster_rows = F, cluster_cols = F, annotation_col = coldata[,"Condition",drop=F], gaps_col = 6, show_colnames = F, show_rownames = F, fontsize = 15, annotation_colors = list("Condition" = c("Adult" = "#F8766D", "Developing" = "#00BFC4")), border_color = NA, scale = "row", main = "Genes with intragenic >6kbp L1HS-L1PA4\nIntersection")

pheatmap(log2(adult_FL_L1HS_L1PA4_location_no_chrun_counts[which(rowSums(adult_FL_L1HS_L1PA4_location_no_chrun_counts) > 0),]+0.5), cluster_rows = F, cluster_cols = F, annotation_col = coldata[,"Condition",drop=F], gaps_col = 6, show_colnames = F, show_rownames = F, fontsize = 15, annotation_colors = list("Condition" = c("Adult" = "#F8766D", "Developing" = "#00BFC4")), border_color = NA, scale = "row", main = "Genes with intragenic >6kbp L1HS-L1PA4\nAdult specific")
# dev.off()
```


### CUT&RUN peaks

How many peaks over FL L1HS-L1PA4 are present in adulthood? How many in development? How many are shared? 
```{r message=FALSE}
files <- list.files(path = c("/Volumes/MyPassport/JGJSeq158_2cutorun/deeptools/CR116_DA735_H3K4me3_S11_L1HS_L1PA2_L1PA3_L1PA4_peaks/",
                             "/Volumes/MyPassport/JGJSeq158_2cutorun/deeptools/CR116_DA736_H3K4me3_S14_L1HS_L1PA2_L1PA3_L1PA4_peaks/",
                             "/Volumes/MyPassport/JGJSeq158_2cutorun/deeptools/CR116_DA737_H3K4me3_S17_L1HS_L1PA2_L1PA3_L1PA4_peaks/"), full.names = T)
adult_L1HS_L1PA2_L1PA3_L1PA4_peaks <- do.call(rbind, lapply(files, fread, header = F))
colnames(adult_L1HS_L1PA2_L1PA3_L1PA4_peaks) <- c("te_chr", "te_start", "te_end", "te_dot", "te_dot1", "te_strand", "te_id", "te_info", "peak_chr", "peak_start", "peak_end", "peak_dot", "peak_dot1", "peak_id", "overlap")
adult_L1HS_L1PA2_L1PA3_L1PA4_peaks <- adult_L1HS_L1PA2_L1PA3_L1PA4_peaks[which(!duplicated(adult_L1HS_L1PA2_L1PA3_L1PA4_peaks[,c("te_chr", "te_start", "te_end", "te_dot", "te_dot1", "te_strand", "te_id", "te_info")])),]

files <- list.files(path = c("/Volumes/MyPassport/JGJSeq169_145cutorun/deeptools/CR118_DA817_H3K4me3_S16_L1HS_L1PA2_L1PA3_L1PA4_peaks/",
                             "/Volumes/MyPassport/JGJSeq169_145cutorun/deeptools/CR119_DA822_H3K4me3_S19_L1HS_L1PA2_L1PA3_L1PA4_peaks/"), full.names = T)
fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks <- do.call(rbind, lapply(files, fread, header = F))
colnames(fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks) <- c("te_chr", "te_start", "te_end", "te_dot", "te_dot1", "te_strand", "te_id", "te_info", "peak_chr", "peak_start", "peak_end", "peak_dot", "peak_dot1", "peak_id", "overlap")
fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks <- fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks[which(!duplicated(fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks[,c("te_chr", "te_start", "te_end", "te_dot", "te_dot1", "te_strand", "te_id", "te_info")])),]

bd_FL_L1HS_L1PA4_peaks <- fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks$te_id
adult_FL_L1HS_L1PA4_peaks <- adult_L1HS_L1PA2_L1PA3_L1PA4_peaks$te_id
bd_adult_FL_L1HS_L1PA4_peaks <- fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks[which(fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks$te_id %in% adult_L1HS_L1PA2_L1PA3_L1PA4_peaks$te_id),]
bd_adult_FL_L1HS_L1PA4_peaks <- bd_adult_FL_L1HS_L1PA4_peaks$te_id
# length(bd_adult_FL_L1HS_L1PA4_peaks)
bd_FL_L1HS_L1PA4_peaks <- bd_FL_L1HS_L1PA4_peaks[which(!bd_FL_L1HS_L1PA4_peaks %in% bd_adult_FL_L1HS_L1PA4_peaks)]
# length(bd_FL_L1HS_L1PA4_peaks)
adult_FL_L1HS_L1PA4_peaks <- adult_FL_L1HS_L1PA4_peaks[which(!adult_FL_L1HS_L1PA4_peaks %in% bd_adult_FL_L1HS_L1PA4_peaks)]
# length(adult_FL_L1HS_L1PA4_peaks)

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/venn_H3K4me3_peaks_FL_L1HS_L1PA2_L1PA3_L1PA4.pdf", width = 16, height = 5)
plot(venneuler(c(Adult = length(adult_FL_L1HS_L1PA4_peaks), # Adult     
                 Development = length(bd_FL_L1HS_L1PA4_peaks), # Brain development
                 "Adult&Development" = length(bd_adult_FL_L1HS_L1PA4_peaks))))
# dev.off()

```

#### Where are these peaks? 

Intragenic vs intergenic 
```{r message=FALSE}
bd_FL_L1HS_L1PA4_peaks <- fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks
adult_FL_L1HS_L1PA4_peaks <- adult_L1HS_L1PA2_L1PA3_L1PA4_peaks
bd_adult_FL_L1HS_L1PA4_peaks <- fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks[which(fetal_L1HS_L1PA2_L1PA3_L1PA4_peaks$te_id %in% adult_L1HS_L1PA2_L1PA3_L1PA4_peaks$te_id), c("te_chr", "te_start", "te_end", "te_id")]
colnames(bd_adult_FL_L1HS_L1PA4_peaks) <- c("chr", "start", "end", "te_id")
# nrow(bd_adult_FL_L1HS_L1PA4_peaks)
bd_FL_L1HS_L1PA4_peaks <- bd_FL_L1HS_L1PA4_peaks[which(!bd_FL_L1HS_L1PA4_peaks$te_id %in% bd_adult_FL_L1HS_L1PA4_peaks$te_id), c("te_chr", "te_start", "te_end", "te_id")]
colnames(bd_FL_L1HS_L1PA4_peaks) <- c("chr", "start", "end", "te_id")
bd_FL_L1HS_L1PA4_peaks$chr <- ifelse(grepl("_", bd_FL_L1HS_L1PA4_peaks$chr), sapply(str_split(bd_FL_L1HS_L1PA4_peaks$chr, "_"), `[[`, 1), bd_FL_L1HS_L1PA4_peaks$chr)
# nrow(bd_FL_L1HS_L1PA4_peaks)
adult_FL_L1HS_L1PA4_peaks <- adult_FL_L1HS_L1PA4_peaks[which(!adult_FL_L1HS_L1PA4_peaks$te_id %in% bd_adult_FL_L1HS_L1PA4_peaks$te_id), c("te_chr", "te_start", "te_end", "te_id")]
colnames(adult_FL_L1HS_L1PA4_peaks) <- c("chr", "start", "end", "te_id")
# nrow(adult_FL_L1HS_L1PA4_peaks)

bd_adult_FL_L1HS_L1PA4_peaks$coords <- add_windows_coords(bd_adult_FL_L1HS_L1PA4_peaks)
bd_FL_L1HS_L1PA4_peaks$coords <- add_windows_coords(bd_FL_L1HS_L1PA4_peaks)
adult_FL_L1HS_L1PA4_peaks$coords <- add_windows_coords(adult_FL_L1HS_L1PA4_peaks)

bd_adult_FL_L1HS_L1PA4_peaks_coords <- bedr.sort.region(bd_adult_FL_L1HS_L1PA4_peaks$coords)
bd_adult_FL_L1HS_L1PA4_peaks_location <- bd_adult_FL_L1HS_L1PA4_peaks[match(bd_adult_FL_L1HS_L1PA4_peaks_coords, bd_adult_FL_L1HS_L1PA4_peaks$coords), c("te_id", "coords")]
bd_adult_FL_L1HS_L1PA4_peaks_location$location <- intergenic_exonic_intronic(bd_adult_FL_L1HS_L1PA4_peaks_location$coords)

bd_FL_L1HS_L1PA4_peaks_coords <- bedr.sort.region(bd_FL_L1HS_L1PA4_peaks$coords)
bd_FL_L1HS_L1PA4_peaks_location <- bd_FL_L1HS_L1PA4_peaks[match(bd_FL_L1HS_L1PA4_peaks_coords, bd_FL_L1HS_L1PA4_peaks$coords), c("te_id", "coords")]
bd_FL_L1HS_L1PA4_peaks_location$location <- intergenic_exonic_intronic(bd_FL_L1HS_L1PA4_peaks_location$coords)

adult_FL_L1HS_L1PA4_peaks_coords <- bedr.sort.region(adult_FL_L1HS_L1PA4_peaks$coords)
adult_FL_L1HS_L1PA4_peaks_location <- adult_FL_L1HS_L1PA4_peaks[match(adult_FL_L1HS_L1PA4_peaks_coords, adult_FL_L1HS_L1PA4_peaks$coords), c("te_id", "coords")]
adult_FL_L1HS_L1PA4_peaks_location$location <- intergenic_exonic_intronic(adult_FL_L1HS_L1PA4_peaks_location$coords)

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/intergenic_exonic_intronic_cut_n_run.pdf", height = 4, width = 3)
ggplot(bd_FL_L1HS_L1PA4_peaks_location, aes(x=location)) + geom_histogram(stat = "count") + theme_classic() + ggtitle("Brain development")
ggplot(bd_adult_FL_L1HS_L1PA4_peaks_location, aes(x=location)) + geom_histogram(stat = "count") + theme_classic() + ggtitle("Intersection")
ggplot(adult_FL_L1HS_L1PA4_peaks_location, aes(x=location)) + geom_histogram(stat = "count") + theme_classic() + ggtitle("Adult")
# dev.off()

```

## Expression of ORF0

We wanted to get some examples of ORF0 usage. For this, we quantified 100bps inside the element and 200bps upstream of it, in antisense of the element. 

Here we just count how many of these regions have some expression in each dataset (from those FL L1HS-PA4 that were found to be expressed in the previous chunks)

We extracted the examples of Supplemental Figure 1c and 3d from these results.
```{r message=FALSE}
path <- "/Volumes/MyPassport/l1_manuscript/TEcounts/unique/"
files <- list.files(path, pattern = "csv")
files <- files[which(!grepl("summary", files))]
files <- files[which(grepl("ORF0", files))]
files <- files[which(startsWith(files, "HuEm") | startsWith(files, "TBI"))]
samples <- unique(sapply(str_split(files, ".FL_L1HS_L1PA4_ORF0_count_matrix_2.csv"), `[[`, 1))

for(i in 1:length(samples)){
  file <- paste(path, files[i], sep="")
  sample <- samples[i]
  
  tmp <- fread(file, data.table=F)
  tmp <- tmp[,c("Geneid", "Strand", "Length", colnames(tmp)[ncol(tmp)])]
  colnames(tmp)[ncol(tmp)] <- sample
    
  if(i == 1){
    ORF0_counts <- tmp
  }else{
    if(all(tmp$Geneid == ORF0_counts$Geneid)){
      ORF0_counts <- cbind(ORF0_counts, tmp[,ncol(tmp),drop=F])  
    }else{
      ORF0_counts <- merge(ORF0_counts, tmp, by=c("Geneid", "Strand", "Length"), all=T)  
    }
  }
  print(sample)
}
rownames(ORF0_counts) <- ORF0_counts$Geneid

ORF0_counts_norm <- ORF0_counts[coldata$sample]

# Normalize by gene expression
ORF0_counts_norm[] <- mapply(ORF0_counts_norm, gene_dds$sizeFactor[coldata$sample], FUN="/")

bd_FL_L1HS_L1PA4_orf0_positive <- data.frame(orf0_positive = apply(ORF0_counts_norm[bd_FL_L1HS_L1PA4_location$id, coldata[which(coldata$Condition == "Developing"),"sample"]] > 0, 2, sum))
bd_FL_L1HS_L1PA4_orf0_positive$sample <- rownames(bd_FL_L1HS_L1PA4_orf0_positive)
bd_adult_FL_L1HS_L1PA4_orf0_positive <- data.frame(orf0_positive = apply(ORF0_counts_norm[bd_adult_FL_L1HS_L1PA4_location$id, coldata$sample] > 0, 2, sum))
bd_adult_FL_L1HS_L1PA4_orf0_positive$sample <- rownames(bd_adult_FL_L1HS_L1PA4_orf0_positive)
adult_FL_L1HS_L1PA4_orf0_positive <- data.frame(orf0_positive = apply(ORF0_counts_norm[adult_FL_L1HS_L1PA4_location$id, coldata[which(coldata$Condition == "Adult"),"sample"]] > 0, 2, sum))
adult_FL_L1HS_L1PA4_orf0_positive$sample <- rownames(adult_FL_L1HS_L1PA4_orf0_positive)

View(ORF0_counts_norm[bd_adult_FL_L1HS_L1PA4_location$id, coldata$sample])

library(ggpubr)
# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/ORF0_positive_elements.pdf", height = 5)
ggarrange(ggplot(adult_FL_L1HS_L1PA4_orf0_positive, aes(x=sample, y=orf0_positive)) + geom_bar(stat="identity", position="dodge") + theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + labs(x="", y="Num of elements with ORF0 expression") + ggtitle(">6kbp L1HS-PA4 expressed elements in adult"),

ggplot(bd_adult_FL_L1HS_L1PA4_orf0_positive, aes(x=sample, y=orf0_positive)) + geom_bar(stat="identity", position="dodge") + theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + labs(x="", y="Num of elements with ORF0 expression") + ggtitle(">6kbp L1HS-PA4 expressed elements in adult and development"),


ggplot(bd_FL_L1HS_L1PA4_orf0_positive, aes(x=sample, y=orf0_positive)) + geom_bar(stat="identity", position="dodge") + theme_classic()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + labs(x="", y="Num of elements with ORF0 expression") + ggtitle(">6kbp L1HS-PA4 expressed elements in development"), ncol = 3)
# dev.off()

```

## Intact vs truncated YY1

Next, we wanted to know if the elements that we found to be expressed (specifically L1HS and L1PA2 which are more likely to be intact 5' UTR) had an intact YY1 binding site. 

To investigate if the YY1 binding site was intact:

1. We wrote bed files with the coordinates of the elements that we found to be expressed in each data set (commented). 
2. We used bedtools getfasta to get the sequence of these elements (commented). 
3. We ran yy1_intact_truncated.py for each of these bed files (and the one with all FL L1HS-PA2 to see if there was an enrichment). 
4. Count how many of these were categorized as truncated or intact.

This section corresponds to Supplemental Figure 8
```{r message=FALSE}
# write.table(TE_bed[bd_FL_L1HS_L1PA4_location$id, c("chr", "start", "end", "dot", "dot", "strand")], file = "/Volumes/MyPassport/l1_manuscript/resources/bd_FL_L1HS_L1PA4_location.bed", quote = F, sep = "\t", row.names = F, col.names = F)
## bedtools getfasta -s -fi /Volumes/MyPassport/annotations/human/gencode/GRCh38.p13/GRCh38.p13.genome.fa -fo bd_FL_L1HS_L1PA4_location.fa  -bed bd_FL_L1HS_L1PA4_location.bed
# write.table(TE_bed[bd_adult_FL_L1HS_L1PA4_location$id, c("chr", "start", "end", "dot", "dot", "strand")], file = "/Volumes/MyPassport/l1_manuscript/resources/bd_adult_FL_L1HS_L1PA4_location.bed", quote = F, sep = "\t", row.names = F, col.names = F)
## bedtools getfasta -s -fi /Volumes/MyPassport/annotations/human/gencode/GRCh38.p13/GRCh38.p13.genome.fa -fo bd_adult_FL_L1HS_L1PA4_location.fa  -bed bd_adult_FL_L1HS_L1PA4_location.bed
# write.table(TE_bed[adult_FL_L1HS_L1PA4_location$id, c("chr", "start", "end", "dot", "dot", "strand")], file = "/Volumes/MyPassport/l1_manuscript/resources/adult_FL_L1HS_L1PA4_location.bed", quote = F, sep = "\t", row.names = F, col.names = F)
## bedtools getfasta -s -fi /Volumes/MyPassport/annotations/human/gencode/GRCh38.p13/GRCh38.p13.genome.fa -fo adult_FL_L1HS_L1PA4_location.fa  -bed adult_FL_L1HS_L1PA4_location.bed 

adult_FL_L1HS_L1PA4_location_intact <- fread("/Volumes/MyPassport/l1_manuscript/resources/adult_FL_L1HS_L1PA4_location_intact.bed", data.table = F)
bd_FL_L1HS_L1PA4_location_intact <- fread("/Volumes/MyPassport/l1_manuscript/resources/bd_FL_L1HS_L1PA4_location_intact.bed", data.table = F)
bd_adult_FL_L1HS_L1PA4_location_intact <- fread("/Volumes/MyPassport/l1_manuscript/resources/bd_adult_FL_L1HS_L1PA4_location_intact.bed", data.table = F)
all_FL_L1HS_L1PA4_location_intact <- fread("/Volumes/MyPassport/l1_manuscript/resources/FL_L1HS_L1PA4_location_intact.bed", data.table = F)

adult_FL_L1HS_L1PA4_location_truncated <- fread("/Volumes/MyPassport/l1_manuscript/resources/adult_FL_L1HS_L1PA4_location_truncated.bed", data.table = F)
bd_FL_L1HS_L1PA4_location_truncated <- fread("/Volumes/MyPassport/l1_manuscript/resources/bd_FL_L1HS_L1PA4_location_truncated.bed", data.table = F)
bd_adult_FL_L1HS_L1PA4_location_truncated <- fread("/Volumes/MyPassport/l1_manuscript/resources/bd_adult_FL_L1HS_L1PA4_location_truncated.bed", data.table = F)
all_FL_L1HS_L1PA4_location_truncated <- fread("/Volumes/MyPassport/l1_manuscript/resources/FL_L1HS_L1PA4_location_truncated.bed", data.table = F)

adult_FL_L1HS_L1PA4_location_intact$section <- "Adult"
bd_FL_L1HS_L1PA4_location_intact$section <- "Fetal"
bd_adult_FL_L1HS_L1PA4_location_intact$section <- "Intersection"
all_FL_L1HS_L1PA4_location_intact$section <- "All"
adult_FL_L1HS_L1PA4_location_intact$type <- "Intact"
bd_FL_L1HS_L1PA4_location_intact$type <- "Intact"
bd_adult_FL_L1HS_L1PA4_location_intact$type <- "Intact"
all_FL_L1HS_L1PA4_location_intact$type <- "Intact"

adult_FL_L1HS_L1PA4_location_truncated$section <- "Adult"
bd_FL_L1HS_L1PA4_location_truncated$section <- "Fetal"
bd_adult_FL_L1HS_L1PA4_location_truncated$section <- "Intersection"
all_FL_L1HS_L1PA4_location_truncated$section <- "All"
adult_FL_L1HS_L1PA4_location_truncated$type <- "Truncated"
bd_FL_L1HS_L1PA4_location_truncated$type <- "Truncated"
bd_adult_FL_L1HS_L1PA4_location_truncated$type <- "Truncated"
all_FL_L1HS_L1PA4_location_truncated$type <- "Truncated"

yy_intact_truncated <- rbind(adult_FL_L1HS_L1PA4_location_intact,
                             bd_FL_L1HS_L1PA4_location_intact,
                             bd_adult_FL_L1HS_L1PA4_location_intact,
                             all_FL_L1HS_L1PA4_location_intact,
                             adult_FL_L1HS_L1PA4_location_truncated,
                             bd_FL_L1HS_L1PA4_location_truncated,
                             bd_adult_FL_L1HS_L1PA4_location_truncated,
                             all_FL_L1HS_L1PA4_location_truncated)

yy_intact_truncated$section <- factor(yy_intact_truncated$section, c("Fetal", "Intersection", "Adult", "All"))

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/yy_intact_truncated_ratio_L1HS_L1PA2.pdf", height = 2, width = 3)
ggplot(yy_intact_truncated_L1HS_L1PA2, aes(x=section, fill=type)) + geom_histogram(stat="count", position = "fill") + #facet_wrap(.~section, scales = "free_y") +
  theme_classic() + labs(y="Count", x="") + ggtitle("YY binding site")
# dev.off()

colnames(yy_intact_truncated) <- c("chr", "start", "end", "dot", "dot2", "strand", "section", "type")
tmp <- merge(yy_intact_truncated, TE_bed[,c("chr", "start", "end", "strand", "id", "TE_subfamily")], by=c("chr", "start", "end", "strand"))
yy_intact_truncated_L1HS_L1PA2 <- tmp[which(tmp$TE_subfamily %in% c("L1HS", "L1PA2")),]

# pdf("/Volumes/MyPassport/FetalCortex/manuscript/January2023/plots/yy_intact_truncated_L1HS_L1PA2.pdf", height = 2, width = 5)
# ggplot(yy_intact_truncated, aes(x=type)) + geom_histogram(stat="count") + facet_wrap(.~section, scales = "free_y") + theme_classic() + labs(y="Count", x="") + ggtitle("YY binding site (L1HS-L1PA4)")
ggplot(yy_intact_truncated_L1HS_L1PA2, aes(x=type)) + geom_histogram(stat="count") + facet_wrap(.~section, scales = "free_y", ncol=4) + theme_classic() + labs(y="Count", x="") + ggtitle("YY binding site (L1HS-L1PA2)")
# dev.off()
```
