---
title: "L1 transcripts and L1 chimeras"
author: "Raquel Garza"
date: "6/24/2023"
output: html_document
---

## Short read bulk RNAseq 

This Rmarkdown refers to the identification and visualization of the L1 transcripts and L1 chimeras as shown in Figure 4. 
```{r}
library(data.table)
library(stringr)
library(openxlsx)
library(DESeq2)
library(ggplot2)
library(ggrepel)
library(VennDiagram)
# Read gtf files
# L1 transcripts (entries are transcripts)
L1transcripts <- fread("/Volumes/MyPassport/l1_manuscript/transcript_assembly/1_mergedsamples/5_transcripts_over1kb_TSS_100bp_chimeric_FL_L1PAs_antisense_flipped.gtf", data.table = F, skip=1)
colnames(L1transcripts) <-  c("chr", "start", "end", "dot1", "dot2", "TE_strand", "TE_id", "TE_info", "TE_overlap", "chr", "start", "end", "dot3", "strand", "dot4", "gene_info")

# Don't taking in consideration alternative chromosomes and such
L1transcripts <- L1transcripts[which(L1transcripts$chr %in% c(paste("chr", 1:22, sep=""), "chrX", "chrY", "chrM")),]

# Read count matrix of all transcripts
counts <- fread("/Volumes/MyPassport/l1_manuscript/transcript_assembly/1_mergedsamples/transcript_count_matrix.csv", data.table = F)

# Parsing information from GTF 
# Transcript id
L1transcripts$transcript_id <- str_replace_all(string = sapply(str_split(L1transcripts$gene_info, 'transcript_id'), `[[`, 2), pattern = '"|;| ', replacement = "")

# Here I load a GTF file of the transcripts over 1kbp. I will use this to subset the L1 chimeras to only those exons which are part of >1kbp transcripts
transcripts_over1kb <- fread("/Volumes/MyPassport/l1_manuscript/transcript_assembly/1_mergedsamples/1_transcripts_over1kb.gtf", data.table=F, skip=1)
# Parse transcript id
transcripts_over1kb$transcript_id <- str_replace_all(string = sapply(str_split(sapply(str_split(transcripts_over1kb$V9, 'transcript_id'), `[[`, 2), "exon_number"), `[[`, 1), pattern = '"|;| ', replacement = "")
# Subset!
```

## Extracting count matrices from these groups
```{r}
L1transcripts_counts <- counts[which(counts$transcript_id %in% L1transcripts$transcript_id),]
rownames(L1transcripts_counts) <- L1transcripts_counts$transcript_id

# Add TE information to these read count matrices
L1transcripts_counts <- merge(L1transcripts_counts, L1transcripts[,c("TE_strand", "TE_id", "strand", "transcript_id")], by="transcript_id")

```


## Read count normalization

To normalize these counts, we will use the sizeFactors of the gene expression data as output from DESeq2
```{r}
# Reading gene quantification ----
path <- "/Volumes/MyPassport/l1_manuscript/gene_counts/default/"
files <- list.files(path, pattern = "csv")
files <- files[which(!endsWith(files, "summary"))]
files <- files[which(startsWith(files, "HuEm") | startsWith(files, "TBI"))]
samples <- sapply(str_split(files, ".gene_count_matrix_2.csv"), `[[`, 1)
conditions <- ifelse(grepl("HuEmCtx", samples), "Developing", "TBI")

coldata <- data.frame(sample= samples,
           condition=conditions,
           file=files)

for(i in 1:length(samples)){
  file <- paste(path, files[i], sep="")
  sample <- samples[i]
  
  tmp <- fread(file, data.table=F)
  tmp <- tmp[,c(1, ncol(tmp))]
  colnames(tmp)[2] <- sample
  
  if(i == 1){
    counts <- tmp
  }else{
    counts <- merge(counts, tmp, by="Geneid", all=T)
  }
}

rownames(coldata) <- coldata$sample
coldata_dev_tbi <- coldata[which(coldata$condition %in% c("Developing", "TBI")),]
rownames(counts) <- counts$Geneid

# Gene DEA ----
gene_dds <- DESeqDataSetFromMatrix(counts[,rownames(coldata_dev_tbi)], coldata_dev_tbi, design = ~ condition)
gene_dds$condition <- relevel(gene_dds$condition, "TBI")
gene_dds <- DESeq(gene_dds)

# Normalize L1 transcripts counts
rownames(L1transcripts_counts) <- make.unique(L1transcripts_counts$transcript_id)
L1transcripts_counts_norm <- L1transcripts_counts[,rownames(coldata)]
L1transcripts_counts_norm[] <- mapply('/', L1transcripts_counts_norm[,names(gene_dds$sizeFactor)], gene_dds$sizeFactor)

```

## L1 transcripts: Venn diagram fetal vs adult

Here we quantify the number of expressed L1 transcripts in the fetal samples vs the adult samples. We output a Venn diagram and Supplemental Table 4.
```{r}
L1transcripts_counts_norm$transcript_id <- rownames(L1transcripts_counts_norm)
L1transcripts_counts_norm_melt <- reshape2::melt(L1transcripts_counts_norm)
L1transcripts_counts_norm_melt <- merge(L1transcripts_counts_norm_melt, coldata[,1:2], by.x="variable", by.y="sample")

L1transcripts_counts_norm_melt_aggr <- aggregate(L1transcripts_counts_norm_melt$value, by=list(L1transcripts_counts_norm_melt$transcript_id, L1transcripts_counts_norm_melt$condition), FUN=sum)

# We consider transcripts to be expressed only if normalized expression is >20
L1transcripts_counts_norm_melt_aggr_expr_dev <- L1transcripts_counts_norm_melt_aggr[which(L1transcripts_counts_norm_melt_aggr$Group.2 == "Developing" & L1transcripts_counts_norm_melt_aggr$x > 20),]

L1transcripts_counts_norm_melt_aggr_expr_adult <- L1transcripts_counts_norm_melt_aggr[which(L1transcripts_counts_norm_melt_aggr$Group.2 == "TBI" & L1transcripts_counts_norm_melt_aggr$x > 20),]


# All expressed transcripts with an antisense L1 in its TSS 
# pdf("/Volumes/MyPassport/FetalCortex/CTG_JGJSeq120_122/transcript_assembly/1_mergedsamples_noGBM/plots/venn_expressed_L1_alt_promoter_transcripts.pdf")
grid.newpage()
venn_object <- venn.diagram(
  x = list(L1transcripts_counts_norm_melt_aggr_expr_adult$Group.1, 
           L1transcripts_counts_norm_melt_aggr_expr_dev$Group.1),
  category.names = c("Adult" , "Development"), filename = NULL
)
grid.draw(venn_object)
# dev.off()

# Get the gene names of all transcripts
gene_annotation <- fread("/Volumes/MyPassport/annotations/human/gencode/v38/gencode.v38.annotation.transcripts.bed", data.table=F, skip=1)
L1transcripts_counts_norm_melt_aggr_expr_adult_out <- merge(L1transcripts_counts_norm_melt_aggr_expr_adult, unique(gene_annotation[,c("V8", "V9")]), by.x="Group.1", by.y="V8", all.x=T)[,-2]
colnames(L1transcripts_counts_norm_melt_aggr_expr_adult_out) <- c("transcript_id", "expression (median of ratios)", "gene_name")
L1transcripts_counts_norm_melt_aggr_expr_dev_out <- merge(L1transcripts_counts_norm_melt_aggr_expr_dev, unique(gene_annotation[,c("V8", "V9")]), by.x="Group.1", by.y="V8", all.x=T)[,-2]
colnames(L1transcripts_counts_norm_melt_aggr_expr_dev_out) <- c("transcript_id", "expression (median of ratios)", "gene_name")

colnames(L1transcripts)[1:3] <- c("TE_chr", "TE_start", "TE_end")
L1transcripts_counts_norm_melt_aggr_expr_adult_out <- merge(L1transcripts_counts_norm_melt_aggr_expr_adult_out, L1transcripts[,c("TE_chr", "TE_start", "TE_end", "TE_strand", "TE_id", "transcript_id")], by= "transcript_id")
L1transcripts_counts_norm_melt_aggr_expr_dev_out <- merge(L1transcripts_counts_norm_melt_aggr_expr_dev_out, L1transcripts[,c("TE_chr", "TE_start", "TE_end", "TE_strand", "TE_id", "transcript_id")], by= "transcript_id")

write.xlsx(list(adult=L1transcripts_counts_norm_melt_aggr_expr_adult_out,
                fetal=L1transcripts_counts_norm_melt_aggr_expr_dev_out), "/Volumes/MyPassport/FetalCortex/manuscript/June2023/alt_prom_L1_transcripts.xlsx")

```


