---
title: "L1 chimeras and L1 transcripts over DE genes per cell types"
author: "Raquel Garza"
date: "6/24/2023"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this R markdown, we perform differential gene expression analysis over the different cell types in our single nuclei RNAseq experiments and we correlate that information with the identified L1-transcripts and chimeras (Figure 4 and Supplemental Figure 4).

## Fetal

The first chunk refers to the fetal samples. We first perform differential gene expression analysis and we quantify per cell type how many of the L1 transcripts and chimeras are specifically expressed on a particular cell type.

Please note that this DEA is performed over genes and not transcripts. 
```{r}
library(data.table)
library(ggpubr)
library(ggplot2)
library(Seurat)
library(stringr)
library(openxlsx)
library(UpSetR)

fetal <- readRDS("/Volumes/MyPassport/FetalCortex/30.03.22/3_combinedUMAP_perCluster/fetalcortex_merged_cluster.rds")
coldata_celltypes <- reshape2::melt(list("Early born neurons" = as.character(c(2,4)),
                    "Basal progenitors" = as.character(c(1,5)),
                    "Apical progenitors" = as.character(c(0,3,6,9)),
                    "Interneurons" = as.character(c(7)),
                    "Cajal Retzius" = as.character(c(8)),
                    "Microglia" = as.character(c(10))))
colnames(coldata_celltypes) <- c("cluster", "cellType")

# Add cell type to Seurat metadata
tmp <- data.frame(cluster = fetal$seurat_clusters)
tmp$cellId <- rownames(tmp) 
tmp <- merge(tmp, coldata_celltypes)
rownames(tmp) <- tmp$cellId
fetal <- AddMetaData(fetal, col.name = "cellType", metadata = tmp[,"cellType",drop=F])

celltypes <- names(table(fetal$cellType))
dea_celltypes_fetal <- list()
for(cell in celltypes) dea_celltypes_fetal[[cell]] <- FindMarkers(fetal, group.by = "cellType", ident.1 = cell)
for(cell in celltypes) dea_celltypes_fetal[[cell]]$gene <- rownames(dea_celltypes_fetal[[cell]]) 
for(cell in celltypes) dea_celltypes_fetal[[cell]] <- dea_celltypes_fetal[[cell]][,c("gene", "p_val",	"avg_log2FC",	"pct.1",	"pct.2",	"p_val_adj")] 
for(cell in celltypes) dea_celltypes_fetal[[cell]]$celltype <- cell

# write.xlsx(x = list("Apical progenitors" = dea_celltypes_fetal$`Apical progenitors`,
#                     "Basal progenitors" = dea_celltypes_fetal$`Basal progenitors`,
#                     "Cajal Retzius" = dea_celltypes_fetal$`Cajal Retzius`,
#                     "Early born neurons" = dea_celltypes_fetal$`Early born neurons`,
#                     "Interneurons" = dea_celltypes_fetal$Interneurons,
#                     "Microglia" = dea_celltypes_fetal$Microglia), file = "/Volumes/MyPassport/FetalCortex/manuscript/June2023/dea_fetal_celltypes.xlsx") # This refers to supplemental table 2

# Supplemental table 4
L1transcripts_fetal <- read.xlsx("/Volumes/MyPassport/FetalCortex/manuscript/June2023/Supplemental_Table_4_alt_prom_L1_transcripts.xlsx", sheet = "fetal")
# Supplemental table 3
L1chimeras_fetal <- read.xlsx("/Volumes/MyPassport/FetalCortex/manuscript/June2023/Supplemental_Table_3_chimeric_L1_transcripts.xlsx", sheet = "fetal")

dea_celltypes_fetal_df <- do.call("rbind", dea_celltypes_fetal)
dea_celltypes_fetal_df <- dea_celltypes_fetal_df[which(dea_celltypes_fetal_df$p_val_adj < 0.05 & dea_celltypes_fetal_df$avg_log2FC > 0),]

dea_celltypes_fetal_df_L1chimeras <- split(dea_celltypes_fetal_df[which(dea_celltypes_fetal_df$gene %in% L1chimeras_fetal$gene_name),], dea_celltypes_fetal_df[which(dea_celltypes_fetal_df$gene %in% L1chimeras_fetal$gene_name),"celltype"])

dea_celltypes_fetal_df_L1transcripts <- split(dea_celltypes_fetal_df[which(dea_celltypes_fetal_df$gene %in% L1transcripts_fetal$gene_name),], dea_celltypes_fetal_df[which(dea_celltypes_fetal_df$gene %in% L1transcripts_fetal$gene_name),"celltype"])
```

## Adult

This second chunk refers to the adult samples. It follows the same logic as before and includes the visualization as shown in Supplemental Figure 4 and the information provided in Supplemental Table 5
```{r}
healthy <- readRDS("/Volumes/MyPassport/TBI/13.07.21/3_mergeSamples/healthy_merged_cluster.rds")
coldata_celltypes <- reshape2::melt(list("Excitatory neurons" = as.character(c(0,1,2,9,10,12,15,20)),
                         "OPC" = as.character(c(18,5)),
                         "Oligodendrocytes" = as.character(c(6,14,19)),
                         "Astrocytes" = as.character(c(3,11,13,21)),
                         "Inhibitory neurons" = as.character(c(4,7,16,8)),
                         "Microglia" = as.character(c(17))))

colnames(coldata_celltypes) <- c("cluster", "cellType")
# Add cell type to Seurat metadata
tmp <- data.frame(cluster = healthy$seurat_clusters)
tmp$cellId <- rownames(tmp) 
tmp <- merge(tmp, coldata_celltypes)
rownames(tmp) <- tmp$cellId
healthy <- AddMetaData(healthy, col.name = "cellType", metadata = tmp[,"cellType",drop=F])

celltypes <- names(table(healthy$cellType))
dea_celltypes_adult <- list()
for(cell in celltypes) dea_celltypes_adult[[cell]] <- FindMarkers(healthy, group.by = "cellType", ident.1 = cell)
for(cell in celltypes) dea_celltypes_adult[[cell]]$gene <- rownames(dea_celltypes_adult[[cell]]) 
for(cell in celltypes) dea_celltypes_adult[[cell]]$celltype <- cell
dea_celltypes_adult_df <- do.call("rbind", dea_celltypes_adult)
dea_celltypes_adult_df <- dea_celltypes_adult_df[which(dea_celltypes_adult_df$p_val_adj < 0.05 & dea_celltypes_adult_df$avg_log2FC > 0),]

for(cell in celltypes) dea_celltypes_adult[[cell]] <- dea_celltypes_adult[[cell]][,c("gene", "p_val",	"avg_log2FC",	"pct.1",	"pct.2",	"p_val_adj")]
# write.xlsx(x = list("Astrocytes" = dea_celltypes_adult$Astrocytes,
#                     "Excitatory neurons" = dea_celltypes_adult$`Excitatory neurons`,
#                     "Inhibitory neurons" = dea_celltypes_adult$`Inhibitory neurons`,
#                     "Microglia" = dea_celltypes_adult$`Microglia`,
#                     "Oligodendrocytes" = dea_celltypes_adult$Oligodendrocytes,
#                     "OPC" = dea_celltypes_adult$OPC), file = "/Volumes/MyPassport/FetalCortex/manuscript/June2023/dea_adult_celltypes.xlsx") # Supplemental table 1

# Supplemental table 4
L1transcripts_adult <- read.xlsx("/Volumes/MyPassport/FetalCortex/manuscript/June2023/Supplemental_Table_4_alt_prom_L1_transcripts.xlsx", sheet = "adult")
# Supplemental table 3
L1chimeras_adult <- read.xlsx("/Volumes/MyPassport/FetalCortex/manuscript/June2023/Supplemental_Table_3_chimeric_L1_transcripts.xlsx", sheet = "adult")

dea_celltypes_adult_df_L1chimeras <- split(dea_celltypes_adult_df[which(dea_celltypes_adult_df$gene %in% L1chimeras_adult$gene_name),], dea_celltypes_adult_df[which(dea_celltypes_adult_df$gene %in% L1chimeras_adult$gene_name),"celltype"])

dea_celltypes_adult_df_L1transcripts <- split(dea_celltypes_adult_df[which(dea_celltypes_adult_df$gene %in% L1transcripts_adult$gene_name),], dea_celltypes_adult_df[which(dea_celltypes_adult_df$gene %in% L1transcripts_adult$gene_name),"celltype"])

# Supplemental Figure 4A
upset(fromList(sapply(dea_celltypes_adult_df_L1chimeras, `[[`, "gene")))
upset(fromList(sapply(dea_celltypes_adult_df_L1transcripts, `[[`, "gene")))
upset(fromList(sapply(dea_celltypes_fetal_df_L1chimeras, `[[`, "gene")))
upset(fromList(sapply(dea_celltypes_fetal_df_L1transcripts, `[[`, "gene")))

# write.xlsx(list("Adult celltype DEG L1transcript" = do.call("rbind", dea_celltypes_adult_df_L1transcripts),
# "Adult celltype DEG L1chimera" = do.call("rbind", dea_celltypes_adult_df_L1chimeras),
# "Fetal celltype DEG L1transcript" = do.call("rbind", dea_celltypes_fetal_df_L1transcripts),
# "Fetal celltype DEG L1chimera" = do.call("rbind", dea_celltypes_fetal_df_L1chimeras)),
# "/Volumes/MyPassport/FetalCortex/manuscript/June2023/Supplemental_Table_5_celltype_DE_genes_L1transcripts_L1chimeras.xlsx")
```



